### ISSUE TITLE
Allow for multiple teams

### ISSUE THREAD

**Issue by kfnic** [wlaurance/slack-for-linux#21]
Since slack allows you to be signed into multiple teams in multiple tabs, it should be possible to introduce a "tabbed" view similar to the OS X client. Ideally There would be a list of teams you're signed in to and a facility to switch between the active teams (CTRL-#, or a similar keybinding?).

**Comment by twolfson** [comment]
I have found screenshots for reference:

![Screenshot](http://a4.mzstatic.com/us/r30/Purple3/v4/88/86/9b/88869b89-cde5-5a50-26cc-3fcd181e8806/screen800x500.jpeg)

**Comment by twolfson** [comment]
@kfnic Is the sidebar always present to switch between teams? If not, when does it appear/disappear?

**Comment by kfnic** [comment]
I believe it's always present once you've signed into a second team.

**Comment by twolfson** [comment]
Cool. I think the next step would be to determine how feasible that detection is for us. e.g. Can we scrape cookies, responses, that switch screen at a fresh login?

I will poke around by the end of next weekend but if anyone else wants to look sooner, let me know.

**Comment by twolfson** [comment]
Yep, it should be possible. We can leverage `TS.getAllTeams()` to get the currently logged in teams. It has a single item per team so we can enable the sidebar when it's greater than 1.

http://viewsource.in/https://slack.global.ssl.fastly.net/31913/js/rollup-core_required_1419975817.js#L11612-11623

The simplest implementation would be to open each of them in an `iframe` and toggle between them. This way we could still listen for notification events.

@kfnic Is there any alternative status displayed when you are viewing a different team window (e.g. busy/away)?

**Comment by kfnic** [comment]
Just tested this and it appears that when you're signed in to two teams you appear online in both, no matter which one is currently in focus.

**Comment by twolfson** [comment]
Sweet! It should be as simple as toggling between iframe visibilities then. I am busy this weekend but will try to get a proof of concept by the end of next weekend. Thanks for all your help @kfnic!

**Comment by twolfson** [comment]
Ah, sorry. I slipped on my promise. I was busy implementing retina support in `spritesmith` all weekend. I will get to this by the end of next weekend.

**Comment by yograterol** [comment]
+1

**Comment by twolfson** [comment]
Started on this tonight but not going to get done. Note to self: Team icons seem to be visible for other teams but not active team in `TS.getAllTeams()`. Thankfully we can leverage `TS.model.team` for this other data.

http://viewsource.in/https://slack.global.ssl.fastly.net/31913/js/rollup-core_required_1419975817.js#L11513

```js
  setUpModel: function(c) {
    var k = !TS.model.ms_logged_in_once;
    TS.model.team = c.team;
    TS.model.bots_legacy = c.team.bots;
    TS.model.team.url = c.url;
    if (!TS.model.last_team_name) {
      TS.model.last_team_name = TS.model.team.name;
      TS.model.last_team_domain = TS.model.team.domain
    }
```

**Comment by twolfson** [comment]
Another note: I couldn't find a nice event emitter to bind to (and even then, it wouldn't be guaranteed to exist when requesting). As a result, I am using a `setTimeout` loop to detect when the page is loaded.

I initially thought of using the `loading` class on `body`

http://viewsource.in/https://slack.global.ssl.fastly.net/31971/js/rollup-client_1420067921.js#L233-234

```js
      $("#loading_welcome").transition({
        opacity: 0
      }, 150, function() {
        $("#col_channels_bg").css("z-index", 0);
        $("#loading-zone").transition({
          opacity: 0
        }, 250, function() {
          $("body").removeClass("loading");
          $("#loading-zone").css({
            opacity: 1,
            left: 0,
            background: "white"
          });
```

but I am going for the feature we want itself:

```js
				// If we haven't completed loading yet, wait for 100ms
				if (!win.TS || !win.TS.getAllTeams || !win.TS.getAllTeams()) {
					return setTimeout(waitForLoaded, 100);
				}

**Comment by twolfson** [comment]
![selection_114](https://cloud.githubusercontent.com/assets/902488/6936168/b0a87f5e-d814-11e4-93c9-1b5966bf7308.png)

Proof of concept seems to work. Click on a link and it updates the iframe URL. But we need more functionality:

- Open other teams in iframes to have appropriate online behavior
- Use same styling as Slack (hopefully reusing their CSS)
    - Use team icons
    - Highlight active team

Then to keep the PRs simple, we should have a follow up PR for:

- Add notification badges on top of team icons
    - Hopefully we can reuse whatever Slack uses

**Comment by twolfson** [comment]
More progress tonight, I am opening multiple iframes and toggling between them.

I remembered another follow up PR:

- Add shortcuts

**Comment by twolfson** [comment]
Toggling and styling are done! Now there is some code cleanup and logic handling to be done (e.g. identifying when you log into a new team, not showing when there is only 1 team).

![selection_117](https://cloud.githubusercontent.com/assets/902488/6991868/e633e054-da72-11e4-9854-37b7b4f1d01d.png)

**Comment by twolfson** [comment]
Decided to defer on handling team sign in detection until we get a better framework in place. The PR is opened: #27.

**Comment by twolfson** [comment]
We had a hiccup with notification support and multi team support. There is a new PR #30 to add multi-team support in with the tray notification handling.

**Comment by twolfson** [comment]
This has been added/released in #30. I will open more issues for the mentioned extra behaviors (e.g. notification badges in-app, team shortcuts).
